### DISCUSSION TITLE
feat(next/build): add prebuild and postbuild hooks

### DISCUSSION THREAD

**Discussion by ctjlewis** [vercel/next.js#39552] [Ideas]
### Describe the feature you'd like to request

[A user on StackOverflow](https://stackoverflow.com/questions/73333837/can-i-intercept-nextjs-compile-step) wants to generate some code from his source files *before* the dev build begins. This is a request to add a feature for providing `next/build` with prebuild and postbuild hooks.

It is pretty common for other build tools to offer prebuild/postbuild hooks, and I imagine it would be relatively easy to plug into the underlying Webpack logic. In effect this is really just two `await` calls sandwiched around the existing build and should be relatively high benefit-to-cost ratio to implement.

### Describe the solution you'd like

The simplest solution is just to add config options:

```js
// next.config.js
{
  buildHooks: {
    preBuild: async () => { /* ... */ },
    postBuild: async () => { /* ... */ },
  },
}
```

And in `next/build/index.ts`:

```ts
export default async function build(
  dir: string,
  conf = null,
  reactProductionProfiling = false,
  debugOutput = false,
  runLint = true
): Promise<void> {
  try {
    /* ... */
    const buildResult = await nextBuildSpan.traceAsyncFn(async () => {
      /* Existing dotenv, config, etc. setup */

      await config.buildHooks.preBuild();
      /* Existing build logic */
      await config.buildHooks.postBuild();
    });
  } catch { /* ... */ }
}
```

### Describe alternatives you've considered

Very limited semi-workaround:

> You can use `nodemon` or something similar to listen to `.next/server/pages/**` and do something *after* the build script, or listen to `src/pages/**` and do something *at the same time* as the dev rebuild, but you cannot get in front of it.

**Comment by CalebBarnes** [comment]
This would be amazing.

**Comment by tatethurston** [comment]
I'd love this for https://github.com/tatethurston/nextjs-routes

**Comment by Nexuist** [comment]
This would be very useful for building next plugins, i.e. custom linting and AST transforming.

**Comment by theoephraim** [comment]
I'd recommend taking a look at astro for an example of a very flexible plugin system - see https://docs.astro.build/en/reference/integrations-reference/

The lack of flexibility with next makes it very difficult to build extensions that enhance the next.js core.

**Comment by usulpro** [comment]
Hey everyone! Would love to have this feature, what is the status now?

**Comment by staticshock** [comment]
This would also be useful for firebase deployments (at least in their current state.) `firebase-tools` is their [fairly inflexible] CLI that calls `next build` under the hood when deploying Next.js [via Firebase Hosting](https://firebase.google.com/docs/hosting/frameworks/nextjs). The tooling has a number of limitations, and Next.js build hooks, if they existed, would have allowed me to counter some of those limitations.

**Comment by techdragon** [comment]
Its kind of crazy this _isnt_ already built in... I found it extremely surprising.

**Comment by ValentinGurkov** [comment]
Would love to see that as well
